documentation/development-docs/design-guidelines/piping-best-practices.md

```diff
@@ -0,0 +1,136 @@
+
+```powershell
+# --- Piping scenario ---
+# Setting and removing an individual object
+Get-AzTopLevelResource -ResourceGroupName "MyResourceGroup" -Name "MyTopLevelResource" | Set-AzTopLevelResource @additionalParams
+Get-AzTopLevelResource -ResourceGroupName "MyResourceGroup" -Name "MyTopLevelResource" | Remove-AzTopLevelResource
+
+# Setting and removing a collection of objects
+Get-AzTopLevelResource | Set-AzTopLevelResource @additionalParams
+Get-AzTopLevelResource | Remove-AzTopLevelResource
+
+
+# --- Non-piping scenario ---
+# Setting and removing an individual object
+$resource = Get-AzTopLevelResource -ResourceGroupName "MyResourceGroup" -Name "MyTopLevelResource"
+Set-AzTopLevelResource -InputObject $resource @additionalParams
+Remove-AzTopLevelResource -InputObject $resource
+
+# Setting and removing a collection of objects
+Get-AzTopLevelResource | ForEach-Object { Set-AzTopLevelResource -InputObject $_ @additionalParams }
+Get-AzTopLevelResource | ForEach-Object { Remove-AzTopLevelResource -InputObject $_ }
+```
+
+Another time that this scenario applies is when you have cmdlets for child resources that need information about the parent (top-level) resource. For example you can pipe in the whole parent object to the `New-AzChildResource` and `Get-AzChildResource` cmdlets to get the child resources, and then pipe the child resource object to the `Remove-AzChildResource` and `Set-AzChildResource` cmdlets.
+
+```powershell
+# --- Piping scenario ---
+# Getting all of child resources from all of the parent resources and removing them
+Get-AzTopLevelResource | Get-AzChildResource | Remove-AzChildResource
+
+# Getting all of the child resources from all of the parent resources in a resource group and removing them
+Get-AzTopLevelResource -ResourceGroupName "MyResourceGroup" | Get-AzChildResource | Remove-AzChildResource
+
+# Getting all of the child resources from a specific parent resource and removing them
+Get-AzTopLevelResource -ResourceGroupName "MyResourceGroup" -Name "MyTopLevelResource" | Get-AzChildResource | Remove-AzChildResource
+
+
+# --- Non-piping scenario ---
+# Getting all of the child resources from a specific parent resource and removing them
+$resource = Get-AzTopLevelResource -ResourceGroupName "MyResourceGroup" -Name "MyTopLevelResource"
+$childResource = Get-AzChildResource -InputObject $resource
+Remove-AzChildResource -InputObject $childResource
+```
+
+### Using the `-ResourceId` Parameter
+
+#### Short explanation
+For all resources, `-ResourceId` should be implemented for the `Remove-*`, `Set-*` and `Update-*` cmdlets (and any other cmdlet where an existing resource is being operated on, such as `Start-*`, `Stop-*`, etc.). The implementation of this will be a new parameter set, like the following:
+
+```
+Remove-AzTopLevelResource -ResourceId <string> [...] [-PassThru] [-WhatIf] [-Confirm]
```

Comment:
What about piping in parent resource id?

Reply:
@maddieclayton that shouldn't be a scenario we support since we're piping by property name, and the `-ResourceId` parameter should only ever grab its value from the `ResourceId` property of the resource we are acting upon with the cmdlet.